src/providers/cryptoauthlib/generate_random.rs

```diff
@@ -10,29 +10,45 @@ impl Provider {
         op: psa_generate_random::Operation,
     ) -> Result<psa_generate_random::Result> {
         let mut random_bytes = vec![0u8; 0];
-        // calculate loop count
-        let call_count = (op.size + rust_cryptoauthlib::ATCA_RANDOM_BUFFER_SIZE - 1)
-            / rust_cryptoauthlib::ATCA_RANDOM_BUFFER_SIZE;
-        // loop
-        for _i in 0..call_count {
-            let mut buffer = Vec::with_capacity(rust_cryptoauthlib::ATCA_RANDOM_BUFFER_SIZE);
-            let err = self.device.random(&mut buffer);
-            match err {
-                rust_cryptoauthlib::AtcaStatus::AtcaSuccess => {
-                    // append buffer vector to result vector
-                    random_bytes.append(&mut buffer);
+        let zero_vector = vec![0u8, op.size as u8];
+        let mut loop_count = 0u8;
+        // external loop to retry generation if vector is not secure
+        loop {
+            // calculate internal loop count
+            let call_count = (op.size + rust_cryptoauthlib::ATCA_RANDOM_BUFFER_SIZE - 1)
+                / rust_cryptoauthlib::ATCA_RANDOM_BUFFER_SIZE;
+            // internal loop for vector size greater than buffer size
+            for _i in 0..call_count {
+                let mut buffer = Vec::with_capacity(rust_cryptoauthlib::ATCA_RANDOM_BUFFER_SIZE);
+                let err = self.device.random(&mut buffer);
```

Comment:
Hmm, I just realised something - since a bigger request (i.e. one asking for more than `ATCA_RANDOM_BUFFER_SIZE`) can end up calling this `random` more than once, is it possible that one individual call returns all zeroes, but the others return something else? I'm just wondering if the loop checking for zeroes should instead be after each `random` call, instead of the whole combined thing.

Reply:
"is it possible that one individual call returns all zeroes, but the others return something else?"

Yes, every other call of `random` should return different result so it is near impossible to get vector `ATCA_RANDOM_BUFFER_SIZE` * 2 of zeros. This is why I think checking only one time at the end if returned vector is all zeros is needed. Also I think when returned vector has at least one non zero value then is no longer considered "unsafe".

Reply:
Hi @akazimierskigl - I am substituting for @ionut-arm here because he is now on leave until the new year. I am slightly lacking some context, so forgive me if I'm misunderstanding. To resolve this question I think I need to understand the underlying cause of a vector of zeroes from the back-end HRNG. Is it purely a probabilistic thing? Is a sequence of 32 zeroes as likely as any other result? If so, then I'm not sure why it needs any special handling, because you might just as well worry about about any other specific set of values. However, if the source of the zeroes is due to some sort of known transient failure or unready state in the HRNG, then I would agree with Ionut that we really should be checking it on every call to the hardware, because otherwise it is compromising the overall entropy of the result. I hope this makes sense.

Reply:
"Is it purely a probabilistic thing? Is a sequence of 32 zeroes as likely as any other result?"
Yes.
"However, if the source of the zeroes is due to some sort of known transient failure or unready state in the HRNG"
It's not.
I've added handling in a first place only because it was required by cipher algorithm to use random non-zero initialization vector.

Reply:
Okay, thanks. In that case it strikes me that we're worrying over nothing, really. It's a vanishingly-unlikely case. A single check on the collected final result is probably good enough, and for all practical purposes that check would never fail anyway.

Reply:
Thanks @paulhowardarm, @akazimierskigl ! I misunderstood what the initial cause for the checks for 0 was, that's a lesson in communication for me :) I thought they'd result from an error in the CAL chip. Shall review once again, apologies for the massive delay this has caused.